packages/tangle/schedulerutils/bufferqueue.go

```diff
@@ -73,21 +69,76 @@ func (b *BufferQueue) Submit(msg Element, rep float64) error {
 		nodeQueue = NewNodeQueue(nodeID)
 	}
 
-	if float64(nodeQueue.Size()+size)/rep > b.maxQueue {
-		return ErrInboxExceeded
-	}
+	// first we submit the message, and if it turns out that the node doesn't have enough bandwidth to submit, it will be removed by dropHead
 	if !nodeQueue.Submit(msg) {
 		panic("message already submitted")
 	}
-
 	// if the node was not active before, add it now
 	if !nodeActive {
 		b.activeNode[nodeID] = b.ringInsert(nodeQueue)
 	}
-	b.size += size
+	b.size += msg.Size()
+
+	// if max buffer size exceeded, drop from head of the longest mana-scaled queue
+	if b.size > b.maxBuffer {
+		return b.dropHead(accessManaRetriever)
+	}
+
 	return nil
 }
 
+func (b *BufferQueue) dropHead(accessManaRetriever func(identity.ID) float64) (messagesDropped []ElementID) {
+	start := b.Current()
+
+	// remove as many messages as necessary to stay within max buffer size
+	for b.Size() > b.maxBuffer {
+		// find longest mana-scaled queue
+		maxScale := math.Inf(-1)
+		var maxNodeID identity.ID
+		for q := start; ; {
+			nodeMana := accessManaRetriever(q.NodeID())
+			if nodeMana > 0.0 {
+				if scale := float64(q.Size()) / nodeMana; scale > maxScale {
+					maxScale = scale
+					maxNodeID = q.NodeID()
+				}
+			} else if q.Size() > 0 {
+				maxScale = math.Inf(1)
+				maxNodeID = q.NodeID()
+			}
+			q = b.Next()
+			if q == start {
+				break
+			}
+		}
+		longestQueue := b.activeNode[maxNodeID].Value.(*NodeQueue)
+		// find oldest submitted and not-ready message in the longest queue
+		var oldestMessage Element
+
+		for _, v := range longestQueue.submitted {
+			if oldestMessage == nil || oldestMessage.IssuingTime().After((*v).IssuingTime()) {
+				oldestMessage = *v
+			}
+		}
+
+		// if the submitted message is older than the oldest ready message, then drop the submitted message
+		// otherwise drop the oldest ready message
```

Comment:
```suggestion
		// if the oldest not-ready message is older than the oldest ready message, drop the former otherwise the latter
```
Improved the comment